add dark theme

cuppajoeman opened the pull request:
Aside from the dark theme this commit adds a button which lets you toggle the color scheme which saves your setting between sessions.

eric-wieser commented:
#deploy

eric-wieser commented:
Hmm, [that crashed](https://github.com/leanprover-community/doc-gen/runs/5105797765?check_suite_focus=true)...

eric-wieser commented:
#deploy

eric-wieser commented:
@gebner, any idea what's causing deployment failures? Is this the "runner is too puny" problem that motivated self-hosted runners elsewhere?

ericrbg commented:
hmm, I doubt this is what it is, but you need to tell the main `.py` to copy stuff over for it to work, see my changes to `print_docs.py` [here](https://github.com/leanprover-community/doc-gen/pull/153/files)

gebner commented:
> any idea what's causing deployment failures? Is this the "runner is too puny" problem that motivated self-hosted runners elsewhere?

Yes, I only moved the regular CI over to the self-hosted runners but forgot about `#deploy` since that's a separate workflow.

#deploy

gebner commented on detect_color_scheme.js:
```diff
@@ -0,0 +1,16 @@
+
+//dark theme preferred, set document with a `data-theme` attribute
+if (theme=="dark") {
+     document.documentElement.setAttribute("data-theme", "dark");
```
This will only pick up the setting the first time the page is loaded.  If I switch the system setting from light to dark, the docs will still be light.

Maybe the easiest way would be to just integrate https://github.com/GoogleChromeLabs/dark-mode-toggle, which seems to handle this case quite robustly.

cuppajoeman replied:
My newest commit should fix this, but I haven't had a chance to test it.

cuppajoeman replied:
> This will only pick up the setting the first time the page is loaded. If I switch the system setting from light to dark, the docs will still be light.
> 
> Maybe the easiest way would be to just integrate https://github.com/GoogleChromeLabs/dark-mode-toggle, which seems to handle this case quite robustly.

Does the newest commit solve this for you?

gebner commented on style.css:
```diff
@@ -20,10 +29,49 @@ nav { line-height: normal; }
+    --body-bg: white;
+    --code-bg: #f3f3f3;
+    --text-color: black;
+    --text-alt-color: hsl(210, 100%, 30%);
```
```suggestion
    --link-color: hsl(210, 100%, 30%);
```

cuppajoeman replied:
Changed, the variable name in the newest commit.

gebner commented on style.css:
```diff
@@ -20,10 +29,49 @@ nav { line-height: normal; }
+    --body-bg: #333;
+    --code-bg: #363333;
+    --text-color: #eee;
+    --text-alt-color: #58a6ff;
```
I've found the yellow a bit glaring in the screenshot you've posted.  How about something more subdued, like `#fad956` (copied from v8.dev)?

eric-wieser replied:
Note that this is already changed, `#58a6ff` is not yellow

gebner commented:
#deploy

gebner commented:
#deploy

eric-wieser commented on theme_toggler.js:
```diff
@@ -0,0 +1,27 @@
+toggleSwitch.addEventListener('click', switchTheme, false);
+
+// also check to see if the user changes their theme settings while the page is loaded.
+window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', event => {
```
This should probably also either clear or override the local storage

cuppajoeman replied:
Newest changes should fix this.

eric-wieser commented on detect_color_scheme.js:
```diff
@@ -0,0 +1,16 @@
+
+// this changes the value of the css varaibles to make it dark
+if (theme=="dark") {
+     document.documentElement.setAttribute("data-theme", "dark");
```
nit: indent is weird here
```suggestion
    document.documentElement.setAttribute("data-theme", "dark");
```

eric-wieser reviewed (commented):
There are a few leftover light colors, like:
https://github.com/leanprover-community/doc-gen/blob/8b81b1a117a3cfcdc4660a68e8ea423e39a78cf7/style.css#L146-L152

and

https://github.com/leanprover-community/doc-gen/blob/8b81b1a117a3cfcdc4660a68e8ea423e39a78cf7/style.css#L617-L618

and

https://github.com/leanprover-community/doc-gen/blob/8b81b1a117a3cfcdc4660a68e8ea423e39a78cf7/style.css#L629-L630

and

https://github.com/leanprover-community/doc-gen/blob/8b81b1a117a3cfcdc4660a68e8ea423e39a78cf7/style.css#L208

eric-wieser commented on style.css:
```diff
@@ -20,10 +29,49 @@ nav { line-height: normal; }
     --content-width: 55vw;
 }
+
+[data-theme="dark"] {
```
```suggestion
[data-theme="dark"] {
    color-scheme: dark;

```
which changes the scrollbars to dark too

gebner commented:
#deploy

digama0 commented:
There is a black text header at https://leanprover-community.github.io/mathlib_docs_demo/commands.html##explode%20/%20#explode_widget, and the red color of `Prop` in the code block is too dark.

gebner commented:
> Does the newest commit solve this for you?

Not really.  Changing the system style doesn't take effect immediately on the docs website.  Sometimes I also have to press the toggle twice before it takes effect.

Two other things:
 - The toggle button is not visible on portrait (less than 700px wide)
 - The hamburger button only has light colors (also for portrait mode).
 - Between ~700-800px window width, the toggle button overflows into a second line (because it is too wide).
 - The toggle button is inside the search form (that might get sent to google)

What I would suggest for the portrait styling issues is to 1) put the toggle button after the form, 2) remove the header-side-padding, and 3) put an empty <div></div> at the beginning of the header.  I would also change the button text to something shorter and more descriptive, that is, "dark" or "🌓" instead of "toggle theme".

cuppajoeman commented:
I've added some more changes today to add more variables with respect to Eric's post. 

I see that @gebner still has the issue with his operating system color scheme changes not affecting the site, to that end I don't really know why that `window.matchMedia` doesn't work, so if anyone wants to take a look at that please do.

As for the other styling issues that gebner mentioned, I can try to do that as well, but it probably wouldn't happen until the end of the day today or in the next day, so if anyone wants to take that on today then there wouldn't be any conflicts.

cuppajoeman commented:
> 



> > Does the newest commit solve this for you?
> 
> Not really. Changing the system style doesn't take effect immediately on the docs website. Sometimes I also have to press the toggle twice before it takes effect.
> 
> Two other things:
> 
>     * The toggle button is not visible on portrait (less than 700px wide)
> 
>     * The hamburger button only has light colors (also for portrait mode).
> 
>     * Between ~700-800px window width, the toggle button overflows into a second line (because it is too wide).
> 
>     * The toggle button is inside the search form (that might get sent to google)
> 
> 
> What I would suggest for the portrait styling issues is to 1) put the toggle button after the form, 2) remove the header-side-padding, and 3) put an empty `<div></div>` at the beginning of the header. I would also change the button text to something shorter and more descriptive, that is, "dark" or "first_quarter_moon" instead of "toggle theme".

I've added these changes

eric-wieser commented:
#deploy

robertylewis commented:
Some weird-looking spots, on a spotcheck:

![image](https://user-images.githubusercontent.com/4967469/155635817-cdc6c1eb-65a8-4b00-9c78-f55150e6abc1.png)
(same on all the "general documentation" pages)

This appears everywhere except the index page, and clicking it does nothing:
![image](https://user-images.githubusercontent.com/4967469/155635957-e347e23b-6b85-40c2-8399-bc9e46b02cf7.png)

robertylewis commented:
I would also prefer the toggle to be more subtle than a big button at the top. As someone who doesn't use dark themes, I may be blind to this. But it isn't common web design to put this kind of toggle so prominently on a page, is it?

cuppajoeman commented:
> Some weird-looking spots, on a spotcheck:
> 
> ![image](https://user-images.githubusercontent.com/4967469/155635817-cdc6c1eb-65a8-4b00-9c78-f55150e6abc1.png) (same on all the "general documentation" pages)
> 
> This appears everywhere except the index page, and clicking it does nothing: ![image](https://user-images.githubusercontent.com/4967469/155635957-e347e23b-6b85-40c2-8399-bc9e46b02cf7.png)

Thanks for catching these - I'll fix up these things tonight. 

As for your other comment - I'm not really sure whether it's good practice or not to put the button up there, but I've definitely seen websites which do that.

cuppajoeman commented:
When I load up the tactics page it looks like this: 
![image](https://user-images.githubusercontent.com/19198104/155803354-e5a1990e-ee01-4274-aca9-f4267465655c.png)

which is unlike what @robertylewis was reporting, so probably they just need to refresh the page to get the new css (ctrl-shift-r) usually works for that. Besides that I've updated the button position and removed the old toggle button, can we rebuild and see how it looks?

bryangingechen commented:
#deploy

cuppajoeman commented:
I think with the newest changes this seems almost good to go!
